src/Symfony/Component/Translation/Translatable/DateTimeTranslatable.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Translation\Translatable;
+
+use Symfony\Contracts\Translation\TranslatableInterface;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+/**
+ * Wrapper around PHP IntlDateFormatter for date and time
+ * The timezone from the DateTime instance is used instead of the server's timezone.
+ *
+ * Implementation of the ICU recommendation to first format advanced parameters before translation.
+ *
+ * @see https://unicode-org.github.io/icu/userguide/format_parse/messages/#format-the-parameters-separately-recommended
+ *
+ * @author Sylvain Fabre <syl.fabre@gmail.com>
+ */
+class DateTimeTranslatable implements TranslatableInterface
+{
+    private $dateTime;
+    private $dateType;
+    private $timeType;
+
+    private $formatters = [];
```

Comment:
having a separate cache per Translatable instance will not help much IMO. Each of those Translatable object is generally short-lived and not translated a lot. I think you want to make this a static cache instead to share it between instances)